### ISSUE TITLE
iOS crash

### ISSUE THREAD

**Issue by mwarnett** [codenameone/CodenameOne#1572]
I released my CodenameOne app for ios in March 2013 and have been updating it regularly since then. I am now trying to issue the first update since the new VM became mandatory and I am still crashing the iOS VM and cannot release an update in this state.

I am unable to isolate the problem to sample code, When you pull little pieces of code out of my app, everything works, but together the VM crashes.

The basic scenario is this. My app starts opens an SQLite database and displays the results in a list. The user then syncs to a remote server to update the database. The database is closed by the list routine, the sync routine opens it, updates the database and closes it. When the sync finishes the list function opens the database again and displays the list, that is when the VM crashes. Not every time, it does work at times. Often it crashes right away, sometimes I will need to go through this 2 or 3 times before it crashes.

The Apple crash log consistently shows that the thread that crashes has loaded libsqlite3.dylib. I ran the app on my Mac using the generated sources under XCode and it always crashes in the same place trying to close the cursor.

I have attached the full Apple crash log and a screen shot of the crash in the XCode debugger.

OK it appears I cannot upload the crash log here, only images. I will paste the key parts below. Any idea ho to send you the full crash log?

I am really blocked at this point. I don't feel I can release an update to my iOS app when it is crashing in key functionality.

![exc1](https://cloud.githubusercontent.com/assets/14006739/9535303/d670f880-4cd4-11e5-9341-3581304b585d.png)

Exception Type:  EXC_BAD_ACCESS (SIGSEGV)
Exception Subtype: KERN_INVALID_ADDRESS at 0x00000021
Triggered by Thread:  1

Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0:
0   libsystem_kernel.dylib          0x3820949c mach_msg_trap + 20
1   libsystem_kernel.dylib          0x38209290 mach_msg + 36
2   CoreFoundation                  0x297f35ce __CFRunLoopServiceMachPort + 142
3   CoreFoundation                  0x297f1b94 __CFRunLoopRun + 1012
4   CoreFoundation                  0x2973d6cc CFRunLoopRunSpecific + 472
5   CoreFoundation                  0x2973d4de CFRunLoopRunInMode + 102
6   GraphicsServices                0x310db1a4 GSEventRunModal + 132
7   UIKit                           0x2ceef440 UIApplicationMain + 1436
8   AWMain                          0x00736652 0xcb000 + 6731346
9   AWMain                          0x00116be6 0xcb000 + 310246
10  AWMain                          0x0028b834 0xcb000 + 1837108
11  AWMain                          0x0065da02 0xcb000 + 5843458
12  AWMain                          0x0065da46 0xcb000 + 5843526
13  libdyld.dylib                   0x38156aac start + 0

Thread 1 Crashed:
0   libsystem_pthread.dylib         0x3829ae36 _pthread_mutex_lock + 14
1   libsqlite3.dylib                0x37eb0272 sqlite3_finalize + 38
2   AWMain                          0x00114962 0xcb000 + 301410
3   AWMain                          0x00113a3a 0xcb000 + 297530
4   AWMain                          0x0011390c 0xcb000 + 297228
5   AWMain                          0x000d316c 0xcb000 + 33132
6   AWMain                          0x0080a686 0xcb000 + 7599750
7   AWMain                          0x00773090 0xcb000 + 6979728
8   AWMain                          0x00774198 0xcb000 + 6984088
9   AWMain                          0x00774f56 0xcb000 + 6987606
10  AWMain                          0x0080a904 0xcb000 + 7600388
11  libsystem_pthread.dylib         0x3829cddc _pthread_body + 136
12  libsystem_pthread.dylib         0x3829cd4e _pthread_start + 114
13  libsystem_pthread.dylib         0x3829aaf8 thread_start + 4

**Comment by shannah** [comment]
That crash is occurring inside the GC thread. This is a case of SQLite not being thread safe.  What is happening is that you have a cursor that you don't explicitly close.  So it is being closed automatically inside the finalize() method, which happens in the GC thread.   The fix for this is to make sure that you are explicitly calling close on all of your database cursors and not leaving them to the GC thread.  We are looking at this issue to see if there is something to do to make the API more crash proof, but it is a tricky issue to get right.  In any case, you should be able to fix this by just closing your cursors.

**Comment by mwarnett** [comment]
I actually ran into that problem when I was developing the original code back in the days of the old VM. I forget exactly what was happening but I soon realized that SQLite on iOS was very picky about closing cursors so I have been very careful about that all along. I added logging to my SQL code to keep track of things like that. I don't think I am missing any but I will double check.

Thanks

**Comment by mwarnett** [comment]
I activated my SQL logging and reran the test.

It crashed at the same point. At the time of the crash there was one cursor still open, but it was still in use. The thread that created it was still active. Is it possible you were trying to close that cursor?

Here is the log so you can see what the app is doing. It starts by opening the database and displaying the list. All cursors have been closed.

Warning: Error creating LLDB target at path '/Users/mcw/Library/Developer/Xcode/DerivedData/AWMain-hgiztmwloevjnpcioqlcxsohvlyy/Build/Products/Debug-iphonesimulator/AWMain.app'- using an empty LLDB target which can cause slow memory reads from remote devices.
2015-08-28 03:50:29.208 AWMain[11364:475643] [EDT] 0:0:0,413 - Log.getUniqueDeviceId()->5925409181925376
2015-08-28 03:50:29.375 AWMain[11364:475643] [EDT] 0:0:0,663 - AWMain - init
2015-08-28 03:50:30.019 AWMain[11364:475643] [EDT] 0:0:1,307 - RW - Music 3.48
2015-08-28 03:50:30.036 AWMain[11364:475643] [EDT] 0:0:1,324 - AWMain - start
2015-08-28 03:50:30.057 AWMain[11364:475643] [EDT] 0:0:1,345 - _osVer=8.4
2015-08-28 03:50:30.073 AWMain[11364:475643] [EDT] 0:0:1,361 - osMajVer=8
2015-08-28 03:50:31.363 AWMain[11364:475643] [EDT] 0:0:2,651 - Loading database Music.aw3
2015-08-28 03:50:31.574 AWMain[11364:475752] [null] 0:0:2,862 - SQL: SELECT IMGW,IMGH FROM DBCATALOG;
2015-08-28 03:50:31.609 AWMain[11364:475752] [null] 0:0:2,897 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:31.623 AWMain[11364:475752] [null] 0:0:2,911 - Closing cursor...
2015-08-28 03:50:31.625 AWMain[11364:475752] [null] 0:0:2,913 - Cursor closed
2015-08-28 03:50:31.633 AWMain[11364:475752] [null] 0:0:2,921 - Database is OPEN
2015-08-28 03:50:31.859 AWMain[11364:475752] [null] 0:0:3,147 - SQL: SELECT _id,TITLE,ARTIST,FORMAT,PUBLISHER,MY_RATING,RELEASE_DATE,STATUS,IMAGE1_DATA FROM AUDIOWARE ORDER BY ARTIST,TITLE;
2015-08-28 03:50:31.878 AWMain[11364:475752] [null] 0:0:3,166 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:32.766 AWMain[11364:475643] [EDT] 0:0:4,54 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 916;
2015-08-28 03:50:32.767 AWMain[11364:475643] [EDT] 0:0:4,55 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:32.816 AWMain[11364:475643] [EDT] 0:0:4,104 - Closing cursor...
2015-08-28 03:50:32.817 AWMain[11364:475643] [EDT] 0:0:4,105 - Cursor closed
2015-08-28 03:50:32.884 AWMain[11364:475643] [EDT] 0:0:4,172 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1;
2015-08-28 03:50:32.885 AWMain[11364:475643] [EDT] 0:0:4,173 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:32.900 AWMain[11364:475643] [EDT] 0:0:4,188 - Closing cursor...
2015-08-28 03:50:32.901 AWMain[11364:475643] [EDT] 0:0:4,189 - Cursor closed
2015-08-28 03:50:32.907 AWMain[11364:475643] [EDT] 0:0:4,195 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 2;
2015-08-28 03:50:32.908 AWMain[11364:475643] [EDT] 0:0:4,196 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:32.915 AWMain[11364:475643] [EDT] 0:0:4,202 - Closing cursor...
2015-08-28 03:50:32.918 AWMain[11364:475643] [EDT] 0:0:4,206 - Cursor closed
2015-08-28 03:50:32.919 AWMain[11364:475643] [EDT] 0:0:4,207 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1088;
2015-08-28 03:50:32.920 AWMain[11364:475643] [EDT] 0:0:4,208 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:32.930 AWMain[11364:475643] [EDT] 0:0:4,218 - Closing cursor...
2015-08-28 03:50:32.931 AWMain[11364:475643] [EDT] 0:0:4,219 - Cursor closed
2015-08-28 03:50:32.932 AWMain[11364:475643] [EDT] 0:0:4,220 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1791;
2015-08-28 03:50:32.933 AWMain[11364:475643] [EDT] 0:0:4,221 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:32.939 AWMain[11364:475643] [EDT] 0:0:4,227 - Closing cursor...
2015-08-28 03:50:32.941 AWMain[11364:475643] [EDT] 0:0:4,229 - Cursor closed
2015-08-28 03:50:32.993 AWMain[11364:475643] [EDT] 0:0:4,281 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1660;
2015-08-28 03:50:32.994 AWMain[11364:475643] [EDT] 0:0:4,282 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:33.002 AWMain[11364:475643] [EDT] 0:0:4,290 - Closing cursor...
2015-08-28 03:50:33.002 AWMain[11364:475643] [EDT] 0:0:4,290 - Cursor closed
2015-08-28 03:50:33.004 AWMain[11364:475643] [EDT] 0:0:4,292 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1669;
2015-08-28 03:50:33.006 AWMain[11364:475643] [EDT] 0:0:4,294 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:33.016 AWMain[11364:475643] [EDT] 0:0:4,304 - Closing cursor...
2015-08-28 03:50:33.017 AWMain[11364:475643] [EDT] 0:0:4,305 - Cursor closed
2015-08-28 03:50:33.019 AWMain[11364:475643] [EDT] 0:0:4,307 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1837;
2015-08-28 03:50:33.020 AWMain[11364:475643] [EDT] 0:0:4,308 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:33.025 AWMain[11364:475643] [EDT] 0:0:4,313 - Closing cursor...
2015-08-28 03:50:33.026 AWMain[11364:475643] [EDT] 0:0:4,314 - Cursor closed
2015-08-28 03:50:33.029 AWMain[11364:475643] [EDT] 0:0:4,317 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 569;
2015-08-28 03:50:33.030 AWMain[11364:475643] [EDT] 0:0:4,318 - SQL Operation Complete, Cursor returned
2015-08-28 03:50:33.049 AWMain[11364:475643] [EDT] 0:0:4,337 - Closing cursor...
2015-08-28 03:50:33.051 AWMain[11364:475643] [EDT] 0:0:4,339 - Cursor closed
2015-08-28 03:50:34.674 AWMain[11364:475752] [null] 0:0:5,962 - Closing cursor...
2015-08-28 03:50:34.680 AWMain[11364:475752] [null] 0:0:5,968 - Cursor closed
2015-08-28 03:50:34.681 AWMain[11364:475752] [null] 0:0:5,969 - Database ready
2015-08-28 03:50:34.769 AWMain[11364:475752] Deleting thread 3

At this point my app is running and the user sees a screen of data. The user decides to sync the database with the remote server. In this case there is nothing to sync but we go through the process and update the database with the latest date etc. Again all cursors are closed.

2015-08-28 03:51:06.974 AWMain[11364:475643] [EDT] 0:0:38,262 - Database is CLOSED
2015-08-28 03:51:09.495 AWMain[11364:475965] [null] 0:0:40,783 - Connection to hobo...
2015-08-28 03:51:09.526 AWMain[11364:475965] [null] 0:0:40,814 - SENT: PING
2015-08-28 03:51:10.500 AWMain[11364:475965] [null] 0:0:41,787 - RECEIVED: OK
2015-08-28 03:51:10.501 AWMain[11364:475965] [null] 0:0:41,789 - SENT: QDBNAME=aw
2015-08-28 03:51:10.754 AWMain[11364:475965] [null] 0:0:42,42 - RECEIVED: Music.aw3.data
2015-08-28 03:51:10.800 AWMain[11364:475965] [null] 0:0:42,88 - SENT: QDBDATE
2015-08-28 03:51:10.836 AWMain[11364:475965] [null] 0:0:42,124 - RECEIVED: 2015-08-28
2015-08-28 03:51:10.838 AWMain[11364:475965] [null] 0:0:42,126 - SENT: QDEVRES=5
2015-08-28 03:51:10.871 AWMain[11364:475965] [null] 0:0:42,159 - RECEIVED: 96
2015-08-28 03:51:10.899 AWMain[11364:475965] [null] 0:0:42,187 - SENT: QUPDDBOK=2015-08-27=TITLE,ARTIST,RUNNING_TIME,UPC,SOUND,PLAYED_COUNT,CATALOG_NUMBER,ROWKEY,FORMAT,PUBLISHER,RELEASE_DATE,COPYRIGHT_DATE,MY_RATING,ITEM_CONDITION,COVER_CONDITION,CATEGORY1,CATEGORY2,CATEGORY3,KEYWORDS,PRODUCT_INFO,MY_COMMENTS,FAVORITE,STATUS,USER2,USER3,DISK_COUNT,TRACKS,IMAGE1_DATA
2015-08-28 03:51:10.934 AWMain[11364:475965] [null] 0:0:42,222 - RECEIVED: OK
2015-08-28 03:51:10.935 AWMain[11364:475965] [null] 0:0:42,223 - SENT: GETUCOL
2015-08-28 03:51:10.982 AWMain[11364:475965] [null] 0:0:42,270 - SENT: UPDDB
2015-08-28 03:51:11.449 AWMain[11364:475965] [null] 0:0:42,737 - RECEIVED: OK=0
2015-08-28 03:51:11.451 AWMain[11364:475965] [null] 0:0:42,739 - SENT: GETROW
2015-08-28 03:51:11.486 AWMain[11364:475965] [null] 0:0:42,773 - SQL: SELECT IMGW,IMGH FROM DBCATALOG;
2015-08-28 03:51:11.488 AWMain[11364:475965] [null] 0:0:42,776 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:11.490 AWMain[11364:475965] [null] 0:0:42,778 - Closing cursor...
2015-08-28 03:51:11.492 AWMain[11364:475965] [null] 0:0:42,780 - Cursor closed
2015-08-28 03:51:11.495 AWMain[11364:475965] [null] 0:0:42,782 - Database is OPEN
2015-08-28 03:51:11.496 AWMain[11364:475965] [null] 0:0:42,784 - SENT: QROWKEYS
2015-08-28 03:51:11.580 AWMain[11364:475965] [null] 0:0:42,868 - SQL: SELECT _id,ROWKEY,IMAGE1_DATA FROM AUDIOWARE ORDER BY ROWKEY;
2015-08-28 03:51:11.581 AWMain[11364:475965] [null] 0:0:42,869 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:11.641 AWMain[11364:475965] [null] 0:0:42,929 - Closing cursor...
2015-08-28 03:51:11.642 AWMain[11364:475965] [null] 0:0:42,930 - Cursor closed
2015-08-28 03:51:11.643 AWMain[11364:475965] [null] 0:0:42,931 - SQL: UPDATE DBCATALOG SET LASTUPD = '2015-08-28';
2015-08-28 03:51:11.674 AWMain[11364:475965] [null] 0:0:42,962 - SQL Operation Complete
2015-08-28 03:51:11.675 AWMain[11364:475965] [null] 0:0:42,963 - Database is CLOSED
2015-08-28 03:51:11.676 AWMain[11364:475965] [null] 0:0:42,964 - SENT: DONE
2015-08-28 03:51:11.709 AWMain[11364:475965] [null] 0:0:42,997 - RECEIVED: OK
2015-08-28 03:51:11.725 AWMain[11364:475965] Deleting thread 4

Once the sync is complete we display the updated database. This is basically a repeat of the first function, it is the same code:

2015-08-28 03:51:16.636 AWMain[11364:475643] [EDT] 0:0:47,923 - Loading database Music.aw3
2015-08-28 03:51:16.640 AWMain[11364:475989] [null] 0:0:47,928 - SQL: SELECT IMGW,IMGH FROM DBCATALOG;
2015-08-28 03:51:16.649 AWMain[11364:475989] [null] 0:0:47,937 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:16.650 AWMain[11364:475989] [null] 0:0:47,938 - Closing cursor...
2015-08-28 03:51:16.652 AWMain[11364:475989] [null] 0:0:47,940 - Cursor closed
2015-08-28 03:51:16.653 AWMain[11364:475989] [null] 0:0:47,941 - Database is OPEN
2015-08-28 03:51:16.671 AWMain[11364:475989] [null] 0:0:47,959 - SQL: SELECT _id,TITLE,ARTIST,FORMAT,PUBLISHER,MY_RATING,RELEASE_DATE,STATUS,IMAGE1_DATA FROM AUDIOWARE ORDER BY ARTIST,TITLE;
2015-08-28 03:51:16.674 AWMain[11364:475989] [null] 0:0:47,962 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:16.924 AWMain[11364:475643] [EDT] 0:0:48,212 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 916;
2015-08-28 03:51:16.926 AWMain[11364:475643] [EDT] 0:0:48,213 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:16.934 AWMain[11364:475643] [EDT] 0:0:48,222 - Closing cursor...
2015-08-28 03:51:16.934 AWMain[11364:475643] [EDT] 0:0:48,222 - Cursor closed
2015-08-28 03:51:16.941 AWMain[11364:475643] [EDT] 0:0:48,229 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1;
2015-08-28 03:51:16.942 AWMain[11364:475643] [EDT] 0:0:48,230 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:16.947 AWMain[11364:475643] [EDT] 0:0:48,235 - Closing cursor...
2015-08-28 03:51:16.948 AWMain[11364:475643] [EDT] 0:0:48,236 - Cursor closed
2015-08-28 03:51:16.949 AWMain[11364:475643] [EDT] 0:0:48,237 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 2;
2015-08-28 03:51:16.950 AWMain[11364:475643] [EDT] 0:0:48,238 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:16.956 AWMain[11364:475643] [EDT] 0:0:48,244 - Closing cursor...
2015-08-28 03:51:16.957 AWMain[11364:475643] [EDT] 0:0:48,245 - Cursor closed
2015-08-28 03:51:16.958 AWMain[11364:475643] [EDT] 0:0:48,246 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1088;
2015-08-28 03:51:16.959 AWMain[11364:475643] [EDT] 0:0:48,247 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:16.965 AWMain[11364:475643] [EDT] 0:0:48,253 - Closing cursor...
2015-08-28 03:51:16.967 AWMain[11364:475643] [EDT] 0:0:48,255 - Cursor closed
2015-08-28 03:51:16.968 AWMain[11364:475643] [EDT] 0:0:48,256 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1791;
2015-08-28 03:51:16.969 AWMain[11364:475643] [EDT] 0:0:48,257 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:16.975 AWMain[11364:475643] [EDT] 0:0:48,263 - Closing cursor...
2015-08-28 03:51:17.008 AWMain[11364:475643] [EDT] 0:0:48,296 - Cursor closed
2015-08-28 03:51:17.014 AWMain[11364:475643] [EDT] 0:0:48,302 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1660;
2015-08-28 03:51:17.015 AWMain[11364:475643] [EDT] 0:0:48,303 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:17.022 AWMain[11364:475643] [EDT] 0:0:48,310 - Closing cursor...
2015-08-28 03:51:17.023 AWMain[11364:475643] [EDT] 0:0:48,311 - Cursor closed
2015-08-28 03:51:17.024 AWMain[11364:475643] [EDT] 0:0:48,312 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1669;
2015-08-28 03:51:17.025 AWMain[11364:475643] [EDT] 0:0:48,313 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:17.032 AWMain[11364:475643] [EDT] 0:0:48,320 - Closing cursor...
2015-08-28 03:51:17.032 AWMain[11364:475643] [EDT] 0:0:48,320 - Cursor closed
2015-08-28 03:51:17.034 AWMain[11364:475643] [EDT] 0:0:48,322 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 1837;
2015-08-28 03:51:17.035 AWMain[11364:475643] [EDT] 0:0:48,323 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:17.040 AWMain[11364:475643] [EDT] 0:0:48,327 - Closing cursor...
2015-08-28 03:51:17.040 AWMain[11364:475643] [EDT] 0:0:48,328 - Cursor closed
2015-08-28 03:51:17.042 AWMain[11364:475643] [EDT] 0:0:48,330 - SQL: SELECT IMAGE1_DATA FROM IMAGES WHERE _id = 569;
2015-08-28 03:51:17.043 AWMain[11364:475643] [EDT] 0:0:48,331 - SQL Operation Complete, Cursor returned
2015-08-28 03:51:17.049 AWMain[11364:475643] [EDT] 0:0:48,337 - Closing cursor...
2015-08-28 03:51:17.050 AWMain[11364:475643] [EDT] 0:0:48,338 - Cursor closed
(lldb)          

If you compare this output to the first you will see that the load has not finished, The load thread is still active and there is one cursor still open.

Is it possible you are trying to free that cursor.

Do you see any problems with the flow, it seems fairly straightforward to me. I am creating background threads to load the database and sync the database, so SQL requests are being issued from different threads.

**Comment by codenameone** [comment]
Accessing SQlite on iOS from more than one thread is prohibited. 
This is a common stumbling block though, I think the proper short term solution should be to hide the SQL implementation and proxy all calls effectively making SQLite on iOS threadsafe. It will slow it down, but I think removing the crashing is probably more important than stability. Steve?

**Comment by codenameone** [comment]
I meant prohibited by Apple...

**Comment by mwarnett** [comment]
Wow, never would have even considered that, been using Java too long.

The same code was working on the old VM, lucky?

There was a situation where I tried to read the database from two different threads. I fixed that. I am still using threads, I have a load thread and a sync thread. They are completely self contained, they open, process and close the database. Only one thread runs at a time. Don't know if that is good enough or not.

Did not help though the app still crashes, the initial load works, the sync works, the second load crashes when GC kicks off and tries to close a cursor. I think there is a problem there.

**Comment by shannah** [comment]
> The same code was working on the old VM, lucky?

Yes.  Race conditions are like this. 

> Did not help though the app still crashes, the initial load works, the sync works, the second load crashes when GC kicks off and tries to close a cursor. I think there is a problem there.

Yes.  That still seems to me like the point to look at.  If the cursor is being closed on the GC, then there must be no references left to that cursor in Java.  You have a lot of stuff going on there. The cursor in question could have been opened at any time, even before the user interaction.  Find out where the cursor originated, and you'll know where is isn't closed.

>  I think the proper short term solution should be to hide the SQL implementation and proxy all calls effectively making SQLite on iOS threadsafe. 

That is one solution.  I'm apprehensive about putting the kid-gloves on the iOS implementation only.  E.g. For comic checklist I solved the multithreaded problem for all platforms by creating a ThreadQueue class that works similar to the EDT and ran all SQL through that queue, ensuring that db calls are both off the EDT and on a single thread.

This is the gist of that class and usage:

https://gist.github.com/shannah/2becc8cf4e23b89ef78f

I seem to recall running into thread issues on Sqlite on both Android and iOS - but iOS was definitely more pronounced.

If we start putting extra layers of thread protection, it might be problematic to other people who have set up threading on their db apps to work in some specific way.

For now, I think it might be best to remove the "close()" call from finalize since this will *always* be on a different thread.  This, of course, opens the possibility of memory leaks, but that seems better than crashes.  We could keep a static list of opened cursors to make it easy for developers to debug opened cursors if they run into memory problems.

Apple claims that you can use sqlite from multiple threads as long as the same database connection is not being used simultaneously on different threads. The same database may not be used from multiple threads if any transactions are pending or any statements remain unfinalised.

**Comment by mwarnett** [comment]
OK, I found my rogue Cursor, once I freed that everything seems to be working.

You can go ahead and close this unless you were planning on seeing if there was a better way to handle this.

Thanks for all your help.

**Comment by codenameone** [comment]
Thanks, I have a local change that will make the JavaSE port write to the system out every time the GC collects an unclosed cursor or database object. This should help in debugging this. Regardless I'll leave this issue open and convert it to an RFE for a more threadsafe SQL solution. 
One option described above is to encapsulate the iOS implementation, another would be to port a Java based SQL solution like hsql. I personally prefer the latter.

**Comment by DigNZ** [comment]
This is a year old but its still open.
I use a native interface "DatabaseFix" that on iOS calls these methods in native code:
`    sqlite3_shutdown();
if (sqlite3_config(SQLITE_CONFIG_SERIALIZED) == SQLITE_OK) {
    NSLog(@"sqlite configured to be threadsafe");
}
sqlite3_initialize();`

This shuts down sqlite, sets it to allow access from multiple threads and restarts it. It certainly has a positive impact on stability, if I start a new project and forget to put it in I notice pretty quickly!

Might be something worth looking at in CN1 itself as it makes the thread-safety setting the same as Android so therefore makes it more crossplatform.

I got this idea from this thread:
http://stackoverflow.com/questions/7795973/setting-sqlite-config-sqlite-config-serialized-returns-sqlite-misuse-on-ios-5

still need to close cursors and do so on try/catch/finally (with another try/catch around the call to close()!!). but that gets it working well.

**Comment by codenameone** [comment]
Thanks, that's an interesting bit of information. We'll look into it.
From my hazy recollection this sounds like something that you can easily merge into our code with a pull request which would probably make things easier to all of us...

**Comment by codenameone** [comment]
We recently committed a threadsafe database implementation which should solve this issue in a cross platform way: https://www.codenameone.com/blog/threadsafe-sqlite.html
